### PR TITLE
Update Caddy to v2.3.0

### LINKED ISSUE #206: Update Caddy to v2
Currently the Caddy image that is used is on v1. Since then v2 has been released which is much more stable. It would be helpful if this could be ported to use the official v2 image.
Also there have been issue with older version of Caddy and newer version of Docker(See https://github.com/stefanprodan/dockprom/issues/203).

I would like to take up the issue and create a PR if approved.

### PR DESCRIPTION
**Pull request by bhumijgupta**
Updated Caddy to v2
Fixes #206
Continuation of #207

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,14 +14,21 @@ Clone this repository on your Docker host, cd into dockprom directory and run co
 git clone https://github.com/stefanprodan/dockprom
 cd dockprom
 
-ADMIN_USER=admin ADMIN_PASSWORD=admin docker-compose up -d
+ADMIN_USER=admin ADMIN_PASSWORD=admin ADMIN_PASSWORD_HASH=JDJhJDE0JE91S1FrN0Z0VEsyWmhrQVpON1VzdHVLSDkyWHdsN0xNbEZYdnNIZm1pb2d1blg4Y09mL0ZP docker-compose up -d
```
Comment (nightah) [inline README.md:17]:
```suggestion
ADMIN_USER=admin ADMIN_PASSWORD=JDJhJDE0JE91S1FrN0Z0VEsyWmhrQVpON1VzdHVLSDkyWHdsN0xNbEZYdnNIZm1pb2d1blg4Y09mL0ZP docker-compose up -d
```

File: caddy/Caddyfile [hunk 2]
```diff
@@ -1,39 +1,25 @@
 :9090 {
-    basicauth / {$ADMIN_USER} {$ADMIN_PASSWORD}
-    proxy / prometheus:9090 {
-            transparent
-        }
-
-    errors stderr
-    tls off
+    basicauth /* {
+        {$ADMIN_USER} {$ADMIN_PASSWORD_HASH}
```
Comment (nightah) [inline caddy/Caddyfile:3]:
```suggestion
        {$ADMIN_USER} {$ADMIN_PASSWORD}
```

File: caddy/Caddyfile [hunk 3]
```diff
@@ -1,39 +1,25 @@
...
 :9093 {
-    basicauth / {$ADMIN_USER} {$ADMIN_PASSWORD}
-    proxy / alertmanager:9093 {
-            transparent
-        }
-
-    errors stderr
-    tls off
+    basicauth /* {
+        {$ADMIN_USER} {$ADMIN_PASSWORD_HASH}
```
Comment (nightah) [inline caddy/Caddyfile:10]:
```suggestion
        {$ADMIN_USER} {$ADMIN_PASSWORD}
```

File: caddy/Caddyfile [hunk 4]
```diff
@@ -1,39 +1,25 @@
...
 :9091 {
-    basicauth / {$ADMIN_USER} {$ADMIN_PASSWORD}
-    proxy / pushgateway:9091 {
-            transparent
-        }
-
-    errors stderr
-    tls off
+    basicauth /* {
+        {$ADMIN_USER} {$ADMIN_PASSWORD_HASH}
```
Comment (nightah) [inline caddy/Caddyfile:17]:
```suggestion
        {$ADMIN_USER} {$ADMIN_PASSWORD}
```

File: docker-compose.yml [hunk 5]
```diff
@@ -127,6 +127,7 @@ services:
     environment:
       - ADMIN_USER=${ADMIN_USER:-admin}
       - ADMIN_PASSWORD=${ADMIN_PASSWORD:-admin}
```
Comment (nightah) [inline docker-compose.yml:129]:
```suggestion
```

File: docker-compose.yml [hunk 6]
```diff
@@ -127,6 +127,7 @@ services:
...
+      - ADMIN_PASSWORD_HASH=${ADMIN_PASSWORD_HASH:-JDJhJDE0JE91S1FrN0Z0VEsyWmhrQVpON1VzdHVLSDkyWHdsN0xNbEZYdnNIZm1pb2d1blg4Y09mL0ZP}
```
Comment (nightah) [inline docker-compose.yml:130]:
```suggestion
      - ADMIN_PASSWORD=${ADMIN_PASSWORD:-JDJhJDE0JE91S1FrN0Z0VEsyWmhrQVpON1VzdHVLSDkyWHdsN0xNbEZYdnNIZm1pb2d1blg4Y09mL0ZP}
```

**Review by nightah — DISMISSED** [review]
Thanks for this @bhumijgupta, I don't see any point in keeping two separate variables as whoever updates will resolve their merge conflicts with need to update their hash.

Let's drop the `ADMIN_HASH_PASSWORD` variable and just move forward with `ADMIN_PASSWORD`.

**Comment by bhumijgupta** [comment]
> Thanks for this @bhumijgupta, I don't see any point in keeping two separate variables as whoever updates will resolve their merge conflicts with need to update their hash.
> 
> Let's drop the `ADMIN_HASH_PASSWORD` variable and just move forward with `ADMIN_PASSWORD`.

The only reason I went ahead with ADMIN_HASH_PASSWORD was that while logging in to the Grafana UI user can then enter their plaintext password rather than having to remember the hashed password

**Comment by nightah** [comment]
Actually that's a really good point that completely slipped my mind.

Can I confirm are you one of the users that is/was affected by the Caddy versioning issue?

If so i'm curious/wondering whether or not just trying to pull the official image for 1.0.5 resolves the issue or not (just means we don't need to worry about updating the password in two places).

**Comment by bhumijgupta** [comment]
In my previous workplace, I faced this issue and internally rewrote the config for Caddy v2 to solve the issue. I did not test with pulling the official image for `1.0.5` . But I suggest it wil sooner or later get deprecated. It's better to make a breaking change now and provide support for longer time (with newer Caddy features).

**Comment by bhumijgupta** [comment]
Gentle ping @nightah

**Comment by DigiDr** [comment]
Just FYI, i tried running this instead of the main branch, but caddy crashes with the same index/slice error others have been reporting on v1.x

**Comment by bhumijgupta** [comment]
@DigiDr Can you please tell which docker version you are using?

**Comment by DigiDr** [comment]
`Docker version 20.10.4, build d3cb89e` @bhumijgupta

**Comment by bhumijgupta** [comment]
@DigiDr I could not reproduce the error you mentioned in v20.10.5 . My docker version is specified below.
```
Client: Docker Engine - Community
 Version:           20.10.5
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        55c4c88
 Built:             Tue Mar  2 20:18:20 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.5
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       363e9a8
  Built:            Tue Mar  2 20:16:15 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

```

**Comment by nightah** [comment]
It still isn't very clear the root cause that this issue is looking to resolve and whether this actually is a genuine fix.

My concern above all is that this will be a breaking change for all existing installations (without user intervention) and it's not clear whether this looks to resolve the issue some users appear to have with the Caddy container.

**Comment by DigiDr** [comment]
Updated docker to match yours @bhumijgupta without any improvement, so we can safely rule that out.

**Comment by DigiDr** [comment]
@bhumijgupta @nightah 

Excuse me for confusing matters, but i was pulling the wrong commit. After figuring out the original bug i went back and looked at this again. 

I can confirm that @bhumijgupta fix with an update to v2 does work.

**Review by nightah — APPROVED** [review]
LGTM.

**Pull request merged**